### PR TITLE
Final landing takeoff

### PR DESCRIPTION
**Pull request by NixonCTChan**
Brings over all the changes from the previous landing takeoff PR's, hopefully this will be able to merge easily.

### CODE REVIEW

File: Autopilot/AttitudeManager/Src/attitudeStateClasses.cpp
Comment (Pop0097) [inline Autopilot/AttitudeManager/Src/attitudeStateClasses.cpp]: newline

File: Autopilot/CMakeLists.txt
Comment (Pop0097) [inline Autopilot/CMakeLists.txt]: I don't think this file exists anymore. Chief got rid of it in his last PR.

File: Autopilot/Inc/SensorFusion.hpp
Comment (Fberbari) [inline Autopilot/Inc/SensorFusion.hpp]: Na, not like this. track is just one of the fields of SFOutput, you'l get it from that structure. The user should never have to call this function. It will just become part of the inside mechanisms of sensor fusion.

To use track, just collect the SFOutput struct from the sensorFusion state and access the track field

File: Autopilot/CMakeLists.txt
Comment (Pop0097) [reply Autopilot/CMakeLists.txt]: Yeah, just checked, it isn't in your branch

File: Autopilot/PathManager/Inc/landingTakeoffManager.hpp
Comment (Fberbari) [inline Autopilot/PathManager/Inc/landingTakeoffManager.hpp]: vector class does not seem to get used in this header file, right ? If that's the case then don't include it in the header in an effort to reduce dependancies. Include it in the cpp instead.

Is the same true with AutoSteer ?

File: Autopilot/Inc/SensorFusion.hpp
Comment (Pop0097) [inline Autopilot/Inc/SensorFusion.hpp]: I would rather you just use PI here since it is more clear :)). 

Also, in SensorFusion.cpp you include <math.h> which contains M_PI. If that doesn't work, you can do what is done in waypointManager.hpp with the #ifndef M_PI and all

File: Autopilot/PathManager/Inc/landingTakeoffManager.hpp
Comment (Fberbari) [inline Autopilot/PathManager/Inc/landingTakeoffManager.hpp]: Is this meant to be a boolean ?

File: Autopilot/Inc/SensorFusion.hpp
Comment (Pop0097) [inline Autopilot/Inc/SensorFusion.hpp]: ->

File: Autopilot/PathManager/Inc/landingTakeoffManager.hpp
Comment (Fberbari) [inline Autopilot/PathManager/Inc/landingTakeoffManager.hpp]: This function and the one above are common to both takeoff and landing, right ? Would it be wise to add another comment seperation section to indicate that it's common ?

File: Autopilot/PathManager/Inc/waypointManager.hpp
Comment (Pop0097) [inline Autopilot/PathManager/Inc/waypointManager.hpp]: If waypoint manager isn't using anything from this file, then no need to include it. 

A way to test for it is by commenting out the include statement and building again lol

File: Autopilot/PathManager/Inc/pathManager.hpp
Comment (Fberbari) [inline Autopilot/PathManager/Inc/pathManager.hpp]: I think this file is not used in the header, right ? If that's the case, just stick it in the appropriate .cpp instead

File: Autopilot/PathManager/Inc/pathDatatypes.hpp
Comment (Pop0097) [inline Autopilot/PathManager/Inc/pathDatatypes.hpp]: Should be doubles

File: Autopilot/PathManager/Inc/waypointManager.hpp
Comment (Fberbari) [inline Autopilot/PathManager/Inc/waypointManager.hpp]: This doesen't look like it belongs here. Why did you decide to add it ?

File: Autopilot/PathManager/Inc/landingTakeoffManager.hpp
Comment (Pop0097) [inline Autopilot/PathManager/Inc/landingTakeoffManager.hpp]: Do you think it would be best to put this enum in pathDatatypes.hpp? That way if anything needs it like pathManager.hpp it doesn't need to access the entire functionality of our landing system

File: Autopilot/PathManager/Inc/pathManager.hpp
Comment (Pop0097) [inline Autopilot/PathManager/Inc/pathManager.hpp]: Refer to comment in landingTakeoffManager.hpp

File: Autopilot/PathManager/Inc/landingTakeoffManager.hpp
Comment (Pop0097) [inline Autopilot/PathManager/Inc/landingTakeoffManager.hpp:35]: Love that you put all the constants at the top!

File: Autopilot/PathManager/Src/waypointManager.cpp
Comment (Fberbari) [inline Autopilot/PathManager/Src/waypointManager.cpp:95]: @Pop0097 this make sense to you ?

File: Autopilot/PathManager/Inc/landingTakeoffManager.hpp
Comment (Pop0097) [inline Autopilot/PathManager/Inc/landingTakeoffManager.hpp]: typo: waypoint data

File: Autopilot/Src/SensorFusion.cpp
Comment (Fberbari) [inline Autopilot/Src/SensorFusion.cpp]: Like I say, not like this. User should not have to call this. Track is just another datapoint bundled in the sensor fusion struct.

File: Autopilot/PathManager/Inc/landingTakeoffManager.hpp
Comment (Pop0097) [inline Autopilot/PathManager/Inc/landingTakeoffManager.hpp:81]: I love that you are using reference parameters here! Using proper C++ conventions lesgooooo

**Review by Fberbari — CHANGES_REQUESTED** [review]
I did this all as one off comments. My b. But yeah, a few things to address, lmk if you think I'm wrong on any of them.

Apologies on missing them on my previous review. Things are much clearer in this one once I was able to know exactly what your'e changing

File: Autopilot/PathManager/Inc/pathManager.hpp
Comment (Pop0097) [inline Autopilot/PathManager/Inc/pathManager.hpp:25]: If you don't mind entertaining me, we could try making these protected. 

```
protected:

         _FlightStage stage;
        bool madeLandingPoints;
        bool madeTakeoffPoints;
        bool isError;
        bool isPackage;

```

This means that only pathManager objects and objects of child classes can access these variables. It adds some encapsulation while also making the variables visible to our states. 

Thoughts on this?

File: Autopilot/PathManager/Src/pathStateClasses.cpp
Comment (Pop0097) [inline Autopilot/PathManager/Src/pathStateClasses.cpp]: A good practice is to always initialize pointers to `nullptr` (Actually, this applies to all variables. I debugged waypoint manager for hours only to find I forgot to initialize an integer). That way they don't point to random garbage and cause undefined behaviour if you are not careful.

I will try to find other places where you didn't initialize them, but maybe doing a quick check on your own for spots like this will be great!

Comment (Pop0097) [reply Autopilot/PathManager/Src/pathStateClasses.cpp]: If you don't want to add the = sign. Just add `{}`. It triggers value initialization, which will set pointers to nullptr and other datatypes to their defaults (ex. integers to 0)

File: Autopilot/PathManager/Src/pathStateClasses.cpp
Comment (Pop0097) [inline Autopilot/PathManager/Src/pathStateClasses.cpp:299]: Happy noises. I see reference parameters

File: Autopilot/PathManager/Src/pathStateClasses.cpp
Comment (Pop0097) [inline Autopilot/PathManager/Src/pathStateClasses.cpp]: I'd recommend adding 10-15 metres in the altitude of currentLocation so it's not at ground level

File: Autopilot/PathManager/Src/landingTakeoffManager.cpp
Comment (Pop0097) [inline Autopilot/PathManager/Src/landingTakeoffManager.cpp:3]: If you defined PI in landingTakeoffManager.hpp, you can do:

```
#ifndef M_PI
#define M_PI 3.14159...
#endif
```

**Review by Pop0097 — COMMENTED** [review]
Nice work! My comments are mostly nit-picky!

File: Autopilot/PathManager/Src/waypointManager.cpp
Comment (Pop0097) [reply Autopilot/PathManager/Src/waypointManager.cpp:95]: Yep. We had agreed on it

File: Autopilot/Inc/SensorFusion.hpp
Comment (NixonCTChan) [reply Autopilot/Inc/SensorFusion.hpp]: so I don't even need this function? I thought we agreed sensorFusion would implement this in the future?

File: Autopilot/PathManager/Inc/landingTakeoffManager.hpp
Comment (NixonCTChan) [reply Autopilot/PathManager/Inc/landingTakeoffManager.hpp]: not for autosteer, the coordturns translation command needs the struct

File: Autopilot/PathManager/Inc/pathManager.hpp
Comment (NixonCTChan) [reply Autopilot/PathManager/Inc/pathManager.hpp]: the _FlightStage struct is used here, but i'm probably gonna move it to pathDatatypes like dhruv said.

File: Autopilot/PathManager/Inc/waypointManager.hpp
Comment (NixonCTChan) [reply Autopilot/PathManager/Inc/waypointManager.hpp]: yup, doesn't belong. Will fix

File: Autopilot/Inc/SensorFusion.hpp
Comment (NixonCTChan) [reply Autopilot/Inc/SensorFusion.hpp]: probably gonna delete this soon, seems like track is gonna be done by sensorfusion and it won't need a function to do that

File: Autopilot/PathManager/Inc/waypointManager.hpp
Comment (NixonCTChan) [reply Autopilot/PathManager/Inc/waypointManager.hpp]: yup fixed

File: Autopilot/PathManager/Inc/pathDatatypes.hpp
Comment (NixonCTChan) [reply Autopilot/PathManager/Inc/pathDatatypes.hpp]: true, fixed!

File: Autopilot/PathManager/Inc/pathManager.hpp
Comment (NixonCTChan) [reply Autopilot/PathManager/Inc/pathManager.hpp:25]: does that mean these variables can't be changed by the states? That's something we need, not just visibility.

File: Autopilot/PathManager/Src/landingTakeoffManager.cpp
Comment (NixonCTChan) [reply Autopilot/PathManager/Src/landingTakeoffManager.cpp:3]: what problem is this solving?

**Review by Pop0097 — COMMENTED** [review]
Looks good to me. Will let @Fberbari take a last look before it's approved!

File: Autopilot/Inc/SensorFusion.hpp
Comment (Fberbari) [reply Autopilot/Inc/SensorFusion.hpp]: Yes we did. But this function won't get exposed to the public. It will live inside the guts of sensorFusion with all of Tony's other stuff. When SF_GenerateNewResult is called by Freertos, one of the things that will happen is trackFinding, which will be done by your code. 

But that's Opaque to the user. All they see is a track value that gets set inside the result struct. 

It's a way of abstracting things so users don't have to worry about the complex stuff that goes on in SensorFusion (and it allows us to change SensorFusion guts in the future without having to modify the calling code and such).

**Review by Fberbari — APPROVED** [review]
I'm satisfied, and I'm approving this!!!!!! 

Wait on Dhruv's final go ahead but after that, merge merge merge!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

**Pull request merged**
